### PR TITLE
Issue 1341 primary class

### LINKED ISSUE #1341: Get the primary class of a compilation unit
Is there anyway to create an instance of ClassOrInterfaceDeclaration from a CompilationUnit?
Today, I'm doing this: `ClassOrInterfaceDeclaration clazz = JavaParser.parse(file).getClassByName(name)`
But to do this, I'm having to get the file name and remove the extension.

What do you guys think about including something like this:
``` java
ClassOrInterfaceDeclaration clazz = JavaParser.parse(file).asClass();
```
or
``` java
CompilationUnit compUnit = JavaParser.parse(file);
ClassOrInterfaceDeclaration clazz = new ClassOrInterfaceDeclaration();
clazz.setInstance(compUnit); // setInstance is just a stupid name here to demonstrate the idea :)
```

### PR DESCRIPTION
**Pull request by matozoid**
Fixes #1341

### CODE REVIEW

File: javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java [hunk 1]
```diff
@@ -465,6 +469,26 @@ public AnnotationDeclaration addAnnotationDeclaration(String name, Modifier... m
...
 
+    /**
+     * @return the name that the primary type in this file should have, according to the filename in {@link Storage#getFileName()}.
+     * Empty if no file information is present (when this compilation unit wasn't parsed from a file.)
+     */
+    public Optional<String> getPrimaryTypeName() {
+        return getStorage()
+                .map(Storage::getFileName)
+                .map(filename -> filename.replace(".java", ""))
+                .map(filename -> filename.replace(".jav", ""));
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java]: Yeah, I don't know...

**Comment by matozoid** [comment]
@cunhazera - can you take a look if this is useful to you?

File: javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java [hunk 2]
```diff
@@ -465,6 +469,26 @@ public AnnotationDeclaration addAnnotationDeclaration(String name, Modifier... m
         return getTypes().stream().filter(type -> type.getNameAsString().equals(enumName) && type instanceof EnumDeclaration).findFirst().map(t -> (EnumDeclaration) t);
     }
 
+    /**
+     * @return the name that the primary type in this file should have, according to the filename in {@link Storage#getFileName()}.
+     * Empty if no file information is present (when this compilation unit wasn't parsed from a file.)
+     */
+    public Optional<String> getPrimaryTypeName() {
```
Comment (cunhazera) [inline javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java:475]: Would be something like this?
``` java
CompilationUnit compU = JavaParser.parse(file);
ClassOrInterfaceDeclaration = compU.getClassByName(compU.getPrimaryTypeName.get());
```

File: javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java (hunk 1 above)
Comment (cunhazera) [reply javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java]: Yeah, why?

**Review by cunhazera — APPROVED** [review]

File: javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java (hunk 1 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java]: Maybe on DOS systems they are limited to extensions of 3 characters

File: javaparser-testing/src/test/resources/com/github/javaparser/storage/PrimaryType.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package com.github.javaparser.storage;
+
+public class PrimaryType {
+    
+}
```
Comment (ftomassetti) [inline javaparser-testing/src/test/resources/com/github/javaparser/storage/PrimaryType.java:5]: Newline :)

This is my favorite type of comments :)

File: javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java (hunk 2 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/ast/CompilationUnit.java:475]: That method has been added too: `getPrimaryType`

File: javaparser-testing/src/test/resources/com/github/javaparser/storage/PrimaryType.java (hunk 3 above)
Comment (matozoid) [reply javaparser-testing/src/test/resources/com/github/javaparser/storage/PrimaryType.java:5]: This PR allows edits from maintainers, so if you consider this so important.... ;-)

**Comment by matozoid** [comment]
@cunhazera @ftomassetti java and jav because it said so in the JLS, but really it should be removing the extension. I'll steal @cunhazera 's code tonight.

**Pull request merged**